### PR TITLE
[export] Initial serialization

### PR DESCRIPTION
**Pull request by angelayi**
Serialization TODOs:
- [x] pytree spec: https://github.com/pytorch/pytorch/pull/102577
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] shape env
- [ ] graph module metadata?

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102125

### CODE REVIEW

File: torch/_export/serialize.py [hunk 1]
```diff
@@ -46,30 +69,441 @@ def _reverse_map(d):
...
+    # dummy_leaves = [0] * spec.num_leaves
+    # tree = torch.utils._pytree.tree_unflatten(dummy_leaves, spec)
+    # _, tree = ex_pytree.tree_flatten(tree)
+    # return tree.to_str()
+    return ""
+    # TODO(angelayi): spec
+    return ""
+
+
+def export_signature(sig) -> GraphSignature:
```
Comment (tugsbayasgalan) [inline torch/_export/serialize.py:160]: We should probably consolidate GraphSignature with ExportMetaData fields.

File: torch/_export/serialize.py [hunk 2]
```diff
@@ -46,30 +69,441 @@ def _reverse_map(d):
...
+            raise SerializeError(f"Serializing {node.target} is not supported")
+
+        self.nodes.append(ex_node)
+
+    def handle_get_attr(self, node):
+        pass
+
+    def export_sym_int_op_inputs(self, args) -> List[NamedArgument]:
+        serialized_args = []
+        args_names = ["a", "b"]
```
Comment (tugsbayasgalan) [inline torch/_export/serialize.py:283]: What is "a" and "b"?

File: torch/_export/serialize.py [hunk 3]
```diff
@@ -46,30 +69,441 @@ def _reverse_map(d):
...
+            inputs=self.inputs,
+            nodes=self.nodes,
+            tensor_values=self.tensor_values,
+            sym_int_values=self.sym_int_values,
+            outputs=self.outputs,
+        )
+
+        buffers = {}
+        parameters = {}
+        for name, buffer in graph_module.named_buffers():
```
Comment (tugsbayasgalan) [inline torch/_export/serialize.py:479]: Can you leave a TODO here saying we will probably move this to ExportedProgram?

File: torch/_export/serialize.py [hunk 4]
```diff
@@ -46,30 +69,441 @@ def _reverse_map(d):
...
+            buffers[name] = export_tensor_meta(buffer)
+        for name, parameter in graph_module.named_parameters():
+            parameters[name] = export_tensor_meta(parameter)
+
+
+        # TODO(angelayi): Graph Module metadata?
+        metadata: Dict[str, str] = {}
```
Comment (tugsbayasgalan) [inline torch/_export/serialize.py:486]: Maybe the Signature should be part of metadata.

File: torch/_export/serialize.py (hunk 1 above)
Comment (angelayi) [reply torch/_export/serialize.py:160]: Yes...this part is all tbd until ExportedProgram is figured out (cc @gmagogsfm)

File: torch/_export/serialize.py (hunk 2 above)
Comment (angelayi) [reply torch/_export/serialize.py:283]: Arbitrary names because those ops don't actually have an arg name :P

File: torch/_export/serialize.py [hunk 5]
```diff
@@ -46,33 +70,440 @@ def _reverse_map(d):
...
+
+def export_operator(target, version) -> Operator:
+    if isinstance(target, str):
+        return Operator(name=target, version=version)
+    elif target in _SYM_INT_OPS:
+        return Operator(name=f"{target.__module__}.{target.__name__}", version=version)
+    elif isinstance(target, torch._ops.HigherOrderOperator):
+        return Operator(name=target.__name__, version=version)
+    else:
+        return Operator(name=str(target), version=version)
```
Comment (larryliu0820) [inline torch/_export/serialize.py:146]: @zhxchen17  do you think we still need per-operator version after we have opset version? I think we need to make a decision on this. My hunch is that we don't need per-operator version anymore.

File: torch/_export/serialize.py [hunk 6]
```diff
@@ -46,33 +70,440 @@ def _reverse_map(d):
...
+
+
+class Serializer:
+    def __init__(self, op_version: int = 0):
+        self.inputs: List[Argument] = []
+        self.outputs: List[Argument] = []
+        self.nodes: List[Node] = []
+        self.tensor_values: Dict[str, TensorValue] = {}
+        self.sym_int_values: Dict[str, SymInt] = {}
+        self.op_version: int = op_version
```
Comment (larryliu0820) [inline torch/_export/serialize.py:217]: oh interesting, so this is actually an operator set version but not per-operator version. Then this should be the constant along the lifecycle of `Serializer` then. I would suggest move this out from `Serializer` for now. In the end it should come from the version table.

File: torch/_export/serialize.py (hunk 5 above)
Comment (zhxchen17) [reply torch/_export/serialize.py]: @larryliu0820 no we don't need per op version now.

File: torch/_export/serialize.py (hunk 6 above)
Comment (zhxchen17) [reply torch/_export/serialize.py]: agree

File: torch/_export/serialize.py [hunk 7]
```diff
@@ -1,21 +1,44 @@
-from typing import Optional, Tuple
+import io
```
Comment (zhxchen17) [inline torch/_export/serialize.py:1]: could you move the entire file under `serde/`

File: torch/_export/serialize.py [hunk 8]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+                graph=graph,
+                signature=export_signature(exported_program.graph_signature),
+                call_spec=export_call_spec(exported_program.call_spec),
+                opset_version=self.opset_version
+            ),
+            export_state_dict(exported_program.state_dict),
+        )
+
+
+def serialize(exported_program: ep.ExportedProgram) -> Tuple[GraphModule, bytes]:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:467]: Q1: Should we pass in a version table here?

Q2: Shouldn't we only return a bytes object here?

File: torch/_export/serialize.py [hunk 9]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+
+        if node.target is torch.set_grad_enabled:
+            # Hack for torch.no_grad support. In the long run this should become
+            # a higher order op but this is fine for now. See [NOTE: nograd support]
+            ex_node = Node(
+                target=export_operator("torch.set_grad_enabled"),
+                inputs=[NamedArgument(name="arg", arg=self.export_input(node.args[0]))],
+                outputs=[],
+                metadata=export_metadata(node),
+            )
```
Comment (zhxchen17) [inline torch/_export/serialize.py:233]: Could we exclude this?

File: torch/_export/serialize.py [hunk 10]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+
+_SYM_INT_OPS = {
+    operator.mul,
+    operator.add,
+    operator.sub,
+    operator.floordiv,
+    operator.mod,
+}
+
+def export_sym_int(s: Union[int, torch.SymInt]) -> SymInt:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:94]: def serialize_sym_int ?

File: torch/_export/serialize.py [hunk 11]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+            return SymInt.create(as_symbol=str(s))
     else:
-        raise ValueError(str(s))
+        raise SerializeError(
+            f"SymInt should be either symbol or int, got `{s}` of type `{type(s)}`"
+        )
 
 
-def _extract_tensor_meta(result: torch.Tensor) -> TensorMeta:
+def export_tensor_meta(t: torch.Tensor) -> TensorMeta:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:108]: def serialize_tensor_meta?

File: torch/_export/serialize.py [hunk 12]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+        requires_grad=t.requires_grad,
+        device=Device(type=t.device.type, index=t.device.index),
+        strides=[export_sym_int(s) for s in t.stride()],
         storage_offset=0,
-        layout=_LAYOUTS[result.layout],
+        layout=_TORCH_TO_SERIALIZE_LAYOUT[t.layout],
     )
 
 
+def export_metadata(node: torch.fx.Node) -> Dict[str, str]:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:123]: def serialize_node_meta?

File: torch/_export/serialize.py [hunk 13]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+    # Need an explicit None check instead of walrus operator, because
+    # module_fqn can be the empty string if the node belongs to the root.
+    # The walrus operator returns False on an empty string :(
+    if module_fqn is not None:
+        ret["module_fqn"] = module_fqn
+    # TODO(angelayi) add nn_module_stack and source_fn
+    return ret
+
+
+def export_operator(target) -> str:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:137]: def serialize_operator?

File: torch/_export/serialize.py [hunk 14]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+        return target
+    elif target in _SYM_INT_OPS:
+        return f"{target.__module__}.{target.__name__}"
+    elif isinstance(target, torch._ops.HigherOrderOperator):
+        return target.__name__
+    else:
+        return str(target)
+
+
+def export_call_spec(call_spec: ep.CallSpec) -> CallSpec:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:148]: same

File: torch/_export/serialize.py [hunk 15]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+    # TODO(angelayi): spec
+    return CallSpec(in_spec="", out_spec="")
+
+
+def export_signature(sig: ep.ExportGraphSignature) -> GraphSignature:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:153]: same

File: torch/_export/serialize.py [hunk 16]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+        inputs_to_buffers=sig.inputs_to_buffers,
+        user_inputs=sig.user_inputs,
+        user_outputs=sig.user_outputs,
+        buffers_to_mutate=sig.buffers_to_mutate,
+        backward_signature=backward_signature,
+    )
+    return graph_signature
+
+
+def export_state_dict(state_dict: Dict[str, Any]) -> bytes:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:174]: same

File: torch/_export/serialize.py [hunk 17]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+                )
+
+        return serialized_args
+
+    def is_sym_int_arg(self, arg) -> bool:
+        return isinstance(arg, int) or (
+            isinstance(arg, torch.fx.Node) and arg.name in self.sym_int_values
+        )
+
+    def export_input(self, arg) -> Argument:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:304]: def serialize_input

File: torch/_export/serialize.py [hunk 18]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+
+
+def serialize(exported_program: ep.ExportedProgram) -> Tuple[GraphModule, bytes]:
+    return Serializer().serialize(exported_program)
+
+
+###############################################################################
+
+
 def convert_fake_tensor_to_tensor_meta(
```
Comment (zhxchen17) [inline torch/_export/serialize.py:474]: do we still need this?

File: torch/_export/serialize.py [hunk 19]
```diff
@@ -102,14 +503,14 @@ def get_shape_env(val) -> Optional[ShapeEnv]:
...
-                torch.Tensor, _extract_tensor_meta, val
+                torch.Tensor, export_tensor_meta, val
             )
             del node.meta["val"]
 
     return ep, shape_env
 
 
-def convert_tensor_meta_to_fake_tensor(ep: ExportedProgram, shape_env: ShapeEnv = None) -> ExportedProgram:
+def convert_tensor_meta_to_fake_tensor(ep: ep.ExportedProgram, shape_env: ShapeEnv = None) -> ep.ExportedProgram:
```
Comment (zhxchen17) [inline torch/_export/serialize.py:513]: same, do we still need this?

**Review by zhxchen17 — COMMENTED** [review]
looks great, lots of questions tho.

File: torch/_export/serialize.py [hunk 20]
```diff
@@ -46,33 +69,411 @@ def _reverse_map(d):
...
+
+
+class Serializer:
+    def __init__(self, opset_version: int = 0):
+        self.inputs: List[Argument] = []
+        self.outputs: List[Argument] = []
+        self.nodes: List[Node] = []
+        self.tensor_values: Dict[str, TensorValue] = {}
+        self.sym_int_values: Dict[str, SymInt] = {}
+        self.opset_version: int = opset_version
```
Comment (larryliu0820) [inline torch/_export/serialize.py:201]: Is it possible to do a `Dict[str, Int]` for opset_version? I think it's clear that we want to serialize an opset version per namespace. cc @avikchaudhuri

Comment (zhxchen17) [reply torch/_export/serialize.py:201]: yeah I think we should pass in a dict.

File: torch/_export/serde/schema.py [hunk 21]
```diff
@@ -154,15 +148,23 @@ class GraphSignature:
...
-    parameters: Dict[str, TensorMeta]
-    metadata: Dict[str, str]
+    opset_version: Dict[str, int]
+
+
+@dataclass
+class ExportedProgram:
+    graph_module: GraphModule
     signature: GraphSignature
+    call_spec: CallSpec
```
Comment (avikchaudhuri) [inline torch/_export/serde/schema.py:170]: Can you describe what the purpose of `call_spec` is? Won't `graph_module` have the `call_spec` mappings burned in?

Is there a standard transform that uses this information?

**Review by avikchaudhuri — APPROVED** [review]
Accepting to unblock, lgtm. See comments though.

File: torch/_export/serde/schema.py [hunk 22]
```diff
@@ -112,15 +112,9 @@ class NamedArgument:
...
-@dataclass
-class Operator:
-    name: str
-    version: Optional[int]
-
-
 @dataclass
 class Node:
-    target: Operator
+    target: str
```
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/_export/serde/schema.py:117]: yup, it'll reside in the toplevel GraphModule

File: torch/_export/serde/schema.py [hunk 23]
```diff
@@ -154,15 +148,23 @@ class GraphSignature:
...
-    backward_signature: Optional[BackwardSignature]
 
 
 @dataclass
 class GraphModule:
     graph: Graph
-    buffers: Dict[str, TensorMeta]
-    parameters: Dict[str, TensorMeta]
-    metadata: Dict[str, str]
+    opset_version: Dict[str, int]
```
[the remark this answers is not in the document]
Comment (larryliu0820) [reply torch/_export/serde/schema.py:163]: My naive approach is to serials a dict directly. It will be a namespace to op set version mapping. Example: {“aten”: 3, “custom_namespace”: 4}. We can change this in the future if needed.

File: torch/_export/serde/schema.py (hunk 21 above)
Comment (angelayi) [reply torch/_export/serde/schema.py:170]: call_spec contains the in_spec/out_spec which is used when we're running the graph module eagerly so that the inputs/output formats match how we would run the original function eagerly.

File: torch/_export/serde/schema.py [hunk 24]
```diff
@@ -143,7 +137,7 @@ class Graph:
 @dataclass
 class BackwardSignature:
     gradients_to_parameters: Dict[str, str]
-    gradients_to_userInputs: Dict[str, str]
+    gradients_to_user_inputs: Dict[str, str]
```
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/_export/serde/schema.py:140]: not sure..copied from the internal thrift file

File: torch/_export/serde/schema.py (hunk 23 above)
Comment (zhxchen17) [inline torch/_export/serde/schema.py:163]: @angelayi I think it makes more sense to me if we put opset_version in ExportedProgram, and put signature in GraphModule. The rationale is to closely mirror the schema in sigmoid.

[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/_export/serde/schema.py:163]: OK, so in any case, no mapping of op names (also str) to versions, yeah? That mapping will exist somewhere else in a table / code?

File: torch/_export/serde/schema.py (hunk 21 above)
Comment (avikchaudhuri) [reply torch/_export/serde/schema.py:170]: Bit confused about why we also then have flatten / unflatten calls in a graph module...

Comment (angelayi) [reply torch/_export/serde/schema.py:170]: Mm, it defines the way we pass inputs and receive outputs to the graph module. Zhengxu stated below that it makes more sense to put the GraphSignature in GraphModule, and I think CallSpec should go with wherever the Signature goes.
